lib/rust/lazy-reader/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+#![feature(test)]
+#![deny(unconditional_recursion)]
+#![warn(missing_copy_implementations)]
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unsafe_code)]
+#![warn(unused_import_braces)]
+
+//! This module exports a reader that is able to process large inputs in constant memory.
+
+pub mod decoder;
+
+use decoder::Decoder;
+
+
+
+// ============
+// === Read ===
+// ============
+
+/// Trait for reading input data into a buffer.
+///
+/// Compared to `std::io::Read` this reader supports multiple input encodings.
+pub trait Read {
+    /// The type of the data in the buffer.
+    type Item;
+
+    /// Fills the buffer and returns amount of elements read.
+    ///
+    /// In case it isn't possible to fill the whole buffer (i.e. if an error like EOF is encountered),
+    /// the buffer will be filled with all the data read before encountering such error.
+    fn read(&mut self, buffer:&mut [Self::Item]) -> usize;
+}
+
+impl<R:std::io::Read> Read for R {
```

Comment:
Please use the standard headings style. See [here](https://github.com/enso-org/enso/blob/main/lib/rust/flexer/src/automata/state.rs#L57) for an example.